cloudstack_plugin/volume.py

```diff
@@ -0,0 +1,142 @@
+########
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# * See the License for the specific language governing permissions and
+# * limitations under the License.
+
+import copy
+from cloudify import ctx
+from cloudify.decorators import operation
+from cloudify.exceptions import NonRecoverableError
+
+from cloudstack_plugin.cloudstack_common import (
+    get_cloud_driver,
+    get_resource_id,
+    CLOUDSTACK_ID_PROPERTY,
+    CLOUDSTACK_NAME_PROPERTY,
+    CLOUDSTACK_TYPE_PROPERTY,
+    COMMON_RUNTIME_PROPERTIES_KEYS,
+    USE_EXTERNAL_RESOURCE_PROPERTY,
+)
+
+VOLUME_CLOUDSTACK_TYPE = 'volume'
+
+RUNTIME_PROPERTIES_KEYS = COMMON_RUNTIME_PROPERTIES_KEYS
+
+
+@operation
+def create(**kwargs):
+    """ Create a volume
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    volume = {}
+
+    ctx.logger.debug('reading volume attributes.')
+    volume.update(copy.deepcopy(ctx.node.properties['volume']))
+
+    if 'name' in volume:
+        volume_name = volume['name']
+    else:
+        raise NonRecoverableError("To create a volume, the name of the volume "
+                                  "is needed")
+
+    if 'size' in volume:
+        volume_size = volume['size']
+    else:
+        raise NonRecoverableError("To create a volume, the size of the volume "
+                                  "is needed")
+
+    existing_vol = volume_exists(cloud_driver, volume_name)
+
+    if not existing_vol:
+
+        volume = cloud_driver.create_volume(name=volume_name,
+                                            size=volume_size)
+
+        ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+        ctx.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+            VOLUME_CLOUDSTACK_TYPE
+
+    elif existing_vol and ctx.node.properties[
+            USE_EXTERNAL_RESOURCE_PROPERTY] is False:
+
+        ctx.logger.warn('Volume with name: {0} already exists'.
+                        format(volume_name))
+
+        return
+
+    elif existing_vol and ctx.node.properties[
+            USE_EXTERNAL_RESOURCE_PROPERTY] is True:
+
+        ctx.logger.info('Using existing volume: {0}'.
+                        format(volume_name))
+
+        volume = get_volume_by_name(cloud_driver, volume_name)
+
+        ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+        ctx.instance.runtime_properties[CLOUDSTACK_NAME_PROPERTY] = volume.name
+        ctx.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+            VOLUME_CLOUDSTACK_TYPE
```

Comment:
This implementation does not seem to follow the proper semantics.

What is should to is the following:

if use_external_resource is True, lookup the resource by the resource_id field, if resource_id is not supplied, raise an error. if the resource does not exist, raise an error. if it does exist, set the needed runtime properties and thats it.

if use_external_resource is False, try creating the volume and apply all the validations for size and name.
if a volume with that name already exists than thats ok, currently we allow creating two volumes with the same name, openstack allows it, and so should we, because the unique identifier of a resource is its id, not name.

please have deeper look at how the openstack plugin handles this.